src/Chapter1.hs

```diff
@@ -554,7 +557,14 @@ value after "=" where the condition is true.
 Casual reminder about adding top-level type signatures for all functions :)
 -}
 
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | (x <= y) && (y <= z) = y
+    | (z <= y) && (y <= x) = y
+    | (y <= x) && (x <= z) = x
+    | (z <= x) && (x <= y) = x
+    | (x <= z) && (z <= y) = z
+    | (y <= z) && (z <= x) = z
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you will need to add another guard – `| otherwise = ...`, to tell the compiler, that your pattern matching is exhaustive 🙂

However, as you rightly noticed, you covered __all__ possible situations with these checks, so if the pattern matching won't work on the first 4 cases, then you can say for sure, that the answer is `z` even without additional checks (as they will definitely succeed):

```haskell
mid x y z
    | (x <= y) && (y <= z) = y
    | (z <= y) && (y <= x) = y
    | (y <= x) && (x <= z) = x
    | (z <= x) && (x <= y) = x
    | otherwise = z
```